--- fs/proc/base.c
+++ fs/proc/base.c
@@ -102,6 +102,9 @@
 #include <linux/cn_proc.h>
 #include <linux/task_integrity.h>
 #include <linux/proca.h>
+#if defined(CONFIG_KSU_SUSFS_SUS_MAP) || defined(CONFIG_KSU_SUSFS_OPEN_REDIRECT)
+#include <linux/susfs_def.h>
+#endif // #if defined(CONFIG_KSU_SUSFS_SUS_MAP) || defined(CONFIG_KSU_SUSFS_OPEN_REDIRECT)
 #include <trace/events/oom.h>
 #include "internal.h"
 #include "fd.h"
